packages/localize/src/program-analysis.ts

```diff
@@ -86,112 +88,250 @@ function extractMsg(
   program: ts.Program,
   node: ts.Node,
   descStack: MsgDesc[]
-): ProgramMessage | ts.Diagnostic | undefined {
+): ResultOrError<ProgramMessage | undefined, ts.Diagnostic> {
   if (!isMsgCall(node, program.getTypeChecker())) {
     // We're not interested.
-    return;
+    return {result: undefined};
   }
-  if (node.arguments.length < 2) {
-    return createDiagnostic(
-      file,
-      node,
-      `Expected at least 2 arguments to msg(), got ${node.arguments.length}`
-    );
+  if (node.arguments.length < 1 || node.arguments.length > 2) {
+    return {
+      error: createDiagnostic(
+        file,
+        node,
+        `Expected 1 or 2 arguments to msg(), got ${node.arguments.length}`
+      ),
+    };
   }
-  const [nameArg, contentsArg] = node.arguments;
-  if (!isStaticString(nameArg) || !nameArg.text) {
-    return createDiagnostic(
-      file,
-      nameArg,
-      `Expected first argument to msg() to be a static and non-empty string`
-    );
+  const [templateArg, optionsArg] = node.arguments;
+
+  const templateResult = extractTemplate(templateArg, file);
+  if (templateResult.error) {
+    return {error: templateResult.error};
   }
-  const name = nameArg.text;
+  const {contents, params, isLitTemplate} = templateResult.result;
 
-  if (isStaticString(contentsArg)) {
-    // E.g. msg('foo', 'bar')
-    return {
+  const optionsResult = extractOptions(optionsArg, file);
+  if (optionsResult.error) {
+    return {error: optionsResult.error};
+  }
+  const options = optionsResult.result;
+  if (options.id === undefined) {
+    // TODO(aomarks) Implement.
+    throw new Error('Not yet implemented: auto ID');
+  }
+  const name = options.id;
+
+  return {
+    result: {
       name,
-      contents: [contentsArg.text],
-      node,
       file,
+      node,
+      contents,
+      params,
+      isLitTemplate,
       descStack: descStack.map((desc) => desc.text),
-      isLitTemplate: false,
+    },
+  };
+}
+
+/**
+ * Analyze the options argument to a msg call.
+ */
+export function extractOptions(
+  node: ts.Node | undefined,
+  file: ts.SourceFile
+): ResultOrError<
+  {id?: string; args?: ts.NodeArray<ts.Expression>},
+  ts.Diagnostic
+> {
+  if (node === undefined) {
+    return {result: {}};
+  }
+  if (!ts.isObjectLiteralExpression(node)) {
+    return {
+      error: createDiagnostic(
+        file,
+        node,
+        `Expected second argument to msg() to be an object`
+      ),
     };
   }
 
-  if (isLitTemplate(contentsArg)) {
-    if (ts.isNoSubstitutionTemplateLiteral(contentsArg.template)) {
-      // E.g. msg('foo', html`bar <b>baz</b>`)
+  let id: string | undefined = undefined;
+  let args: ts.NodeArray<ts.Expression> | undefined = undefined;
+
+  for (const property of node.properties) {
```

Comment:
Does `.properties` include object spread?

Reply:
Yep. Here are all the forms, I added this as a comment too:

```ts
const x = {
  a: 0,     // properties > PropertyAssignment > Identifier
  'b': 0,   // properties > PropertyAssignment > StringLiteral
  ['c']: 0, // properties > PropertyAssignment > ComputedPropertyName
  d,        // properties > ShorthandPropertyAssignment
  ...e,     // properties > SpreadAssignment
};
```

https://astexplorer.net/#/gist/4c8b4f5d6e957befe180b429d58fb2ef/1661e78f3669a9fd66c7aafb35040d38a1191cb1